### DISCUSSION TITLE
How to type a dynamic zod schema

### DISCUSSION THREAD

**Discussion by binajmen** [colinhacks/zod#2039] [Q&A] [answered]
Hi,

I'm using a library called `zodix` (https://github.com/rileytomasek/zodix) that allows to easily assert the existence of params in the URL (in a Remix context):
```ts
export async function loader({ params, request }: LoaderArgs) {
  const { id } = zx.parseParams(params, { id: z.string() });
  ...
};
```

It becomes boilerplate-y when you have many variables to assert:
```ts
  const { lang, organizationId, memberId, taskId } = zx.parseParams(params, {
    lang: z.string(),
    organizationId: z.string(),
    memberId: z.string(),
    taskId: z.string(),
  });
```

I'd like to simplify things by creating an utility function:
```ts
export function assertParams(params: Params, ...keys: string[]) {
  return zx.parseParams(
    params,
    keys.reduce((acc, key) => ({ ...acc, [key]: z.string() })) // <-- how can I type this?
  );
}
```

In order to call it that way:
```ts
  const { lang, organizationId, memberId, taskId } = assertParams(params, "lang", "organizationId", "memberId", "taskId");
```

I can't find a way to type correctly the `reduce` return value so that the return value of `assertParams` works like expected. Is there a way to correctly type a generated schema?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
function keysToZodSchema<
    Schema extends z.ZodTypeAny,
    Key extends string,
> ( schema: Schema, ...keys: Key[] ) {
    return keys.reduce(
        ( acc, key ) => ( { ...acc, [ key ]: schema } ),
        {} as Record<Key, typeof schema>
    )
}

const { lang, organizationId, memberId, taskId } =
    keysToZodSchema( z.string(), 'lang', 'organizationId', 'memberId', 'taskId' )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by binajmen** [reply]
Not really. Here a TS playground link to better illustrate my question:
https://tsplay.dev/WP1zew

But as discussed on Discord, this is perhaps a step too far. I leave it as a challenge if you wish to solve it ;)

**Comment by binajmen** [comment, marked as answer]
As mentioned on Discord
> What's the point? I feel like you're trying to optimize and abstract away something that will just make things more awkward for you. I would stick to just using zx.parseParams directly.

Perhaps I should not do it in the first place...
